DataFormats/simulation/src/GeneratorHeader.cxx

```diff
@@ -0,0 +1,169 @@
+
+GeneratorHeader::GeneratorHeader()
+  : mName("Generator"), mTrackOffset(0), mNumberOfTracks(0), mNumberOfAttempts(0), mInfo()
+{
+  /** default constructor **/
+}
+
+/*****************************************************************/
+
+GeneratorHeader::GeneratorHeader(const std::string& name)
+  : mName(name), mTrackOffset(0), mNumberOfTracks(0), mNumberOfAttempts(0), mInfo()
+{
+  /** constructor **/
+}
+
+/*****************************************************************/
+
+GeneratorHeader::GeneratorHeader(const GeneratorHeader& rhs)
+  : mName(rhs.mName),
+    mTrackOffset(rhs.mTrackOffset),
+    mNumberOfTracks(rhs.mNumberOfTracks),
+    mNumberOfAttempts(rhs.mNumberOfAttempts),
+    mInfo(rhs.mInfo)
+{
+  /** copy constructor **/
+}
+
+/*****************************************************************/
+
+GeneratorHeader& GeneratorHeader::operator=(const GeneratorHeader& rhs)
+{
+  /** operator= **/
+
+  if (this == &rhs)
+    return *this;
+  mName = rhs.mName;
+  mTrackOffset = rhs.mTrackOffset;
+  mNumberOfTracks = rhs.mNumberOfTracks;
+  mNumberOfAttempts = rhs.mNumberOfAttempts;
+  mInfo = rhs.mInfo;
+  return *this;
+}
+
+/*****************************************************************/
+
+GeneratorHeader::~GeneratorHeader() { /** default destructor **/}
+
+/*****************************************************************/
+
+void GeneratorHeader::reset()
+{
+  /** reset **/
+
+  mTrackOffset = 0;
+  mNumberOfTracks = 0;
+  mNumberOfAttempts = 0;
+  for (auto& info : mInfo)
+    info.second->reset();
+}
+
+/*****************************************************************/
+
+void GeneratorHeader::print() const
+{
+  /** print **/
+
+  auto name = getName();
+  auto offset = getTrackOffset();
+  auto ntracks = getNumberOfTracks();
+  std::cout << ">> generator: " << name << " | tracks: " << offset << " -> " << offset + ntracks - 1 << std::endl;
+  for (auto const& info : mInfo)
+    info.second->print();
+}
+
+/*****************************************************************/
+
+CrossSectionInfo* GeneratorHeader::getCrossSectionInfo() const
+{
+  /** get cross-section info **/
+
+  std::string key = CrossSectionInfo::keyName();
+  if (!mInfo.count(key))
+    return nullptr;
+  return dynamic_cast<CrossSectionInfo*>(mInfo.at(key));
+}
+
+/*****************************************************************/
+
+CrossSectionInfo* GeneratorHeader::addCrossSectionInfo()
+{
+  /** add cross-section info **/
+
+  std::string key = CrossSectionInfo::keyName();
+  if (!mInfo.count(key))
+    mInfo[key] = new CrossSectionInfo();
+  return static_cast<CrossSectionInfo*>(mInfo.at(key));
+}
+
+/*****************************************************************/
+
+void GeneratorHeader::removeCrossSectionInfo()
+{
+  /** remove cross-section info **/
+
+  removeGeneratorInfo(CrossSectionInfo::keyName());
+}
+
+/*****************************************************************/
+
+HeavyIonInfo* GeneratorHeader::getHeavyIonInfo() const
```

Comment:
all of these getters could be provided by a single template function to reduce code. Something like this
```C++
template<typename InfoType> InfoType* getInfo() {
    if( !mInfo.count(InfoType::constexpr_key )) {
      return nullptr;  
    }
    // etc.
}
```